### ISSUE TITLE
md-tab combined with ui-router and nested states breaks expected behaviour

### ISSUE THREAD

**Issue by MarijnMensinga** [angular/material#5930]
Dear all,

**TL;DR** We use `<md-tab label="Agents" ui-sref="settings.agent.overview" md-active="$state.includes('settings.agent')"></md-tab>` to select the current tab in our settings screen. When we go from state `settings.agent.overview` to `settings.agent.edit`. The tab stays selected which is nice (the watcher for `$state.includes('settings.agent')` does not change). But when we come from `other.state` and go directly to `settings.agent.edit` the md-active gets triggered and this triggers a click ([code](https://github.com/angular/material/commit/6c53d1127fc1cb7d2a86802e5831642a40ff1e41)) which in turn redirects to the state `settings.agent.overview`, skipping over our intended state. How should we fix this the right way?

**The case:** We're working with nested states and a tab navigation to navigate through those states. The basic idea is that we have a settings page with a couple of categories. One of which is the agents category/tab. When this agents tab is opened an overview page is shown. From this page one can click to open a form to edit the selected agent. This edit form lives under the same agents tab, but in a different state (see our states below in app.js).

**The problem:** From a different location (`other.state`) in our application we would like to direct link to the edit form of one specific agent. We simply do this by changing our state to `settings.agent.edit`. The pages gets redirected as we would suspect. But when it loads the md-tabs in our parent state `settings` will be loaded as well and tries to select/activate the correct tab `md-active="$state.includes('settings.agent')"`. Since `settings.agent.edit` is a subste of `settings.agent` the tab `agents` will be activated. But an activate triggers a click ([commit](https://github.com/angular/material/commit/6c53d1127fc1cb7d2a86802e5831642a40ff1e41)). This click in turn triggers the ui-sref to be executed again which redirects our view to the `settings.agent.overview` state. So our intended state `settings.agent.edit` is never shown to the end user.

**My question:** Based on the code snippets below. What is the best way to solve the described problem? 
* Should we build our own custom click function on the tab which detects state changes to determine whether or not we should trigger a state change? 
* Or should there be an md-tab option `md-selected=""` which does select the tab, but does not trigger a click? (an `md-active` cannot be used alongside the md-selected on the same tab)

**templates/settings.html:**
```html
<md-tabs class="md-primary">
    <md-tab label="Other tab 1" ui-sref="other.state"></md-tab>  
    <md-tab label="Agents" ui-sref="settings.agent.overview" md-active="$state.includes('settings.agent')"></md-tab>
	<md-tab label="Other tab 2" ui-sref="other.state"></md-tab>
</md-tabs>

<ui-view/>
```

**app.js:**
```javascript
$stateProvider
	.state("other.state", {
		templateUrl: 'templates/other.html'
	})
	.state("settings", {
		controller: 'settingsController',
		templateUrl: 'templates/settings.html'
	})
	.state("settings.agent", {
		abstract: true,
		template: '<ui-view/>'
	})
	.state("settings.agent.overview", {
		controller: 'settingsAgentOverviewController',
		templateUrl: 'templates/settings/settingsAgentOverview.html'
	})
	.state("settings.agent.edit", {
		controller: 'settingsAgentEditController',
		params: {agentId: { value: null}},
		templateUrl: 'templates/settings/settingsAgentEdit.html'
	})
```

Thank you in advance!
Marijn

**Comment by bwegrzyn** [comment]
This is a problem in my project as well. Would love to see some solutions to this.

**Comment by rudzikdawid** [comment]
+1

**Comment by epelc** [comment]
This is a duplicate of #5351

The cause is `md-tabs` fire a fake click when they are activated. This causes the ui-sref or any `ng-click` handlers to fire and therefore switch states.

**Comment by Frank3K** [comment]
+1

**Comment by bopm** [comment]
:+1: anyone? It's here for petty long time, still nothing.

**Comment by epelc** [comment]
I think this issue is on backlog because it was duplicated a few times. But the pr is scheduled to be merged for 1.0.6(the next version)

**Comment by bopm** [comment]
@epelc according to docs, 1.0.6 is already here. It's not in releases yet, but docs are based on 1.0.6 for some reason.

**Comment by Frank3K** [comment]
The [CHANGELOG.md](https://github.com/angular/material/blob/9e7937eebd47d1d7158aaed6e4acf4cbfd3e248e/CHANGELOG.md) and [package.json](https://github.com/angular/material/blob/0b12f2f0fa1ea23b1f957d1cccab11178a2cb7ab/package.json) indicate that 1.0.5 is the current version.

**Comment by Frank3K** [comment]
Since #6482 has been pulled today this issue should be fixed.

I have tried a similar setup as @MarijnMensinga, but I can not get it working properly. What I end up with is a working situation, but there is a *very* slow fadeOut / fadeIn animation upon navigating between the tabs. Any tips?

**settings.html**

``` html
<md-tabs id="settings" md-dynamic-height md-no-select-click class="md-primary">
	<md-tab data-ui-sref="app.settings.account" md-active="$state.includes('app.settings.account')">
		<md-tab-label>
			{{ 'ACCOUNT' | translate }}
		</md-tab-label>
		<md-tab-body>
			<md-content>
				<div data-ui-view="account"></div>
			</md-content>
		</md-tab-body>
	</md-tab>
	<md-tab data-ui-sref="app.settings.colors" md-active="$state.includes('app.settings.colors')">
		<md-tab-label>
			{{ 'COLORS' | translate }}
		</md-tab-label>
		<md-tab-body>
			<md-content>
				<div data-ui-view="colors"></div>
			</md-content>
		</md-tab-body>
	</md-tab>
</md-tabs>
```

**routing.js**

``` js
		$stateProvider
    		.state('app.settings', {
				abstract: true,
				url: '/settings',
				views: {
					content: {
						templateUrl: 'settings.html'
					}
				}
			})
			.state('app.settings.account', {
				url: '/account/',
				views: {
					account: {
						templateUrl: 'settings/account.html',
						controller: 'SettingsController'
					}
				}
			})
      ...
```

The `settings/account.html` contains the HTML for the tab.

**Comment by Frank3K** [comment]
@epelc: I think the difference comes from inline vs 'external' tabs. 

At first I did not understand the concept of 'external' tabs, but I think I get it now. You only use the `md-tabs` for displaying the tabs itself. Below that (i.e. outside of `<md-tabs></md-tabs>` you display the tab content, right? Isn't that a little contrary to how the tabs are designed to work?

**Comment by epelc** [comment]
Yes I display them outside of <md-tabs> each <md-tab>(note no s) is used
only for the label/clicker.

No it's not outside how they were designed. However iirc the docs only
briefly mention the external tab method. However they do mention it so they
were designed with it in mind. I'd say the primary benefit of using
external tabs especially with ui-router is that you can make each tabs
content a child state and separate the template/controllers much more
easily.

On Thu, Mar 10, 2016 at 4:36 PM, Frank <notifications@github.com> wrote:

> @epelc <https://github.com/epelc>: I think the difference comes from
> inline vs 'external' tabs.
>
> At first I did not understand the concept of 'external' tabs, but I think
> I get it now. You only use the md-tabs for displaying the tabs itself.
> Below that (i.e. outside of <md-tabs></md-tabs> you display the tab
> content, right? Isn't that a little contrary to how the tabs are designed
> to work?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/material/issues/5930#issuecomment-195059605>.
>



-- 
Thank you,
Ed Pelc

**Comment by Frank3K** [comment]
Good point @epelc, I'll try using external tabs.

**Comment by ThomasBurleson** [comment]
This issue is closed as part of our deprecation effort.
    For details, see our post [Spring Cleaning 2016](https://groups.google.com/forum/#!topic/ngMaterial/3Ip9nd68bfk).

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (530/500 (106%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext3.iad.github.net ([192.30.252.194]:33916 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by ps5.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86_1)
	(envelope-from <noreply@github.com>)
	id 1avscB-000L5x-De
	for greg@biomassiv.es; Thu, 28 Apr 2016 15:32:47 -0500
Date: Thu, 28 Apr 2016 13:32:12 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1461875532;
	bh=p3u8173kryAqxgr4pn75TAbygNdUOrp3cnXJokr1KTU=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=WNbL9CCSw/DY3P3F7/HbmuBu192i080Q4BPY1EWli1GgL7Xrem8Vd4udEY/aVOJCg
	 NBv7d/t0FIOg6QzRRZegTtBo2sm6+l4ETj0u0ekkMDMUmJS/BFY2CeVZAtz9rPc1Ew
	 0DccA1WafeFZCRqlS1iX3SRSQdFCyw4Mek7BdHA0=

**Comment by pakhuta** [comment]
add into your control:
`$scope.$state = $state;`

**Comment by masalinas** [comment]
@Frank3K Did you resolve the problem about the slow fadeOut / fadeIn animation upon navigating between the tabs, I'm using the last mayerial release 1.1.1 with the directive md-no-select-click inside md-tabs but the result is the same, persist this  slow fadeOut / fadeIn animation.

Do you resolve it?

**Comment by Frank3K** [comment]
@masalinas: unfortunately not. I have switched to external tabs (as described above). They do not have an animation...

Did you take a look at the new `md-nav-bar`?

**Comment by masalinas** [comment]
@Frank3K Thanks for the response, do you have any example using external tabs?
Regards

**Comment by masalinas** [comment]
@Frank3K I resolved the flickering problem flickering with a little css hack for the angular tabsheet, you must add this css code in your css style app:

``` javascript
/*resolve animation bug in tabsheet*/
#content > md-tabs {
    -webkit-transition: all 0.0s !important;
    transition: all 0.0s !important;
}
```

**Comment by ThomasBurleson** [comment]
@clshortfuse - should we incorporate ^ css fix into `tabs.scss` ?

**Comment by masalinas** [comment]
Not. I added in my own style sheet of course.

**Comment by masalinas** [comment]
@clshortfuse not, I set in my own css style sheet of course. Your response looks like  better.
